### PR TITLE
Use ICMP Ping to local gateway in the watchdog

### PR DESCRIPTION
**Pull request by posterzh**
**Issue**

- Link: https://github.com/NebraLtd/hm-diag/issues/387
- Summary:
Find out the correct connection status to the local gateway and the internet so that watchdog can work reliably.

**How**
- Use dbus interface to the network manager to get the local gateway IP inside the diagnostics container.
- Use ICMP `ping` method to check the connection status to the local gateway(ex, 192.168.0.1) and public IP(8.8.8.8).

**Screenshots**

**References**
- https://github.com/NebraLtd/hm-diag/issues/343
- [icmplib](https://pypi.org/project/icmplib/)

**Checklist**

- [x] Tests added
- [ ] Cleaned up commit history (rebase!)
- [ ] Documentation added
- [ ] Thought about variable and method names

### CODE REVIEW

File: hw_diag/utilities/network_watchdog.py [hunk 1]
```diff
@@ -84,43 +59,60 @@ def __init__(self):
...
         # Log the watchdog intro
-        self.LOGGER.info("Watchdog is started!")
+        self.LOGGER.info("Watchdog started!")
 
     def __del__(self):
         if hasattr(self, 'temp_dir'):
             self.temp_dir.cleanup()
 
-    def have_internet(self, timeout=10) -> bool:
+    def icmp_ping(self, ip: str) -> bool:
```
[the remark this answers is not in the document]
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:69]: sounds good.

File: hw_diag/utilities/network_watchdog.py [hunk 2]
```diff
@@ -35,46 +29,27 @@ class NetworkWatchdog:
...
-               """
-        if NetworkWatchdog._instance is None:
-            NetworkWatchdog._instance = self
-        else:
-            raise RuntimeError("You cannot create another SingletonGovt class")
-
-        # Save the uptime
-        self.up_time = datetime.now()
+        # Save the watchdog start time
+        self.watchdog_start_time = datetime.now()
```
[the remark this answers is not in the document]
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:45]:
```
import time
time.monotonic()
```
This gives the uptime in seconds and looks better.

[the remark this answers is not in the document]
Comment (posterzh) [reply hw_diag/utilities/network_watchdog.py:45]: [uptime](https://pythonhosted.org/uptime/#module-uptime) library works well on any major platform for getting OS uptime.

File: hw_diag/app.py [hunk 3]
```diff
@@ -59,10 +59,11 @@ def init_scheduled_tasks(app) -> None:
     def run_ship_diagnostics_task():
         perform_hw_diagnostics(ship=True)
 
+    watchdog = NetworkWatchdog()
```
[the remark this answers is not in the document]
Comment (pritamghanghas) [reply hw_diag/app.py:62]: my confusion came the fact that mostly functions belong to class and not instance. But if you are sure that self is referenced in callable, I am good.

File: hw_diag/utilities/network_watchdog.py [hunk 4]
```diff
@@ -133,17 +117,10 @@ def ensure_network_connection(self) -> None:
         self.LOGGER.warning(
             f"Network is not connected! Lost connectivity count={self.lost_count}")
 
-        if self.lost_count >= self.NM_RESTART_THRESHOLD:
-            self.LOGGER.warning(
-                "Reached threshold for nm restart for recovering network.")
-            self.restart_network_manager()
```
Comment (pritamghanghas) [inline hw_diag/utilities/network_watchdog.py:139]: why did we get rid of network manager restart? Wouldn't this push our action threshold to 3 hours for non connectivity.

File: hw_diag/utilities/network_watchdog.py [hunk 5]
```diff
@@ -84,43 +54,57 @@ def __init__(self):
...
+        systemd_proxy = Systemd()
+        network_manager_unit = systemd_proxy.get_unit(DBusIds.NETWORK_MANAGER_UNIT_NAME)
+        nm_restarted = network_manager_unit.wait_restart()
         self.LOGGER.info(f"Network manager restarted: {nm_restarted}")
 
     def ensure_network_connection(self) -> None:
-        self.LOGGER.info("Running the watchdog...")
+        self.LOGGER.info("Ensuring the network connection...")
+
+        up_time = timedelta(seconds=uptime())
```
Comment (pritamghanghas) [inline hw_diag/utilities/network_watchdog.py:106]: nice

File: hw_diag/utilities/network_watchdog.py (hunk 4 above)
Comment (pritamghanghas) [reply hw_diag/utilities/network_watchdog.py:139]: ok, found it. Ignore the comment.

**Review by pritamghanghas — APPROVED** [review]

**Pull request merged**